pkg/kp/rcstore/rc.go

```diff
@@ -0,0 +1,95 @@
+package rcstore
+
+import (
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/pods"
+)
+
+type ReplicationControllerDesires struct {
+	replicasDesired int
+	disabled        bool
+}
+
+type ReplicationControllerData struct {
+	// It is expected that these fields will never be changed after an RC is created.
+	id           string
+	manifest     *pods.Manifest
+	nodeSelector labels.Selector
+	podLabels    labels.Set
+
+	watcher desireWatcher
+
+	// It is expected that these fields can be changed.
+	ReplicationControllerDesires
+}
+
+// ReplicationControllerStore represents an interface for persisting replication controllers to Consul,
+// as well as restoring replication controllers from Consul.
+type ReplicationControllerStore interface {
+	// Create creates a replication controller with the specified manifest and selectors.
+	// The node selector is used to determine what nodes the replication controller may schedule on.
+	// The pod label set is:
+	// 1) applied to every pod the replication controller schedules,
+	// 2) used (in addition to replication controller ID) to determine what pods are already scheduled.
+	Create(manifest pods.Manifest, nodeSelector labels.Selector, podLabels labels.Set) (ReplicationRequest, error)
+
+	// List lists all of the replication controllers.
+	List() ([]ReplicationRequest, error)
+}
+
+type desireWatcher interface {
+	// watch(quitAndAck) watches for any changes to the ReplicationControllerDesires associated with the watcher.
+	// watch() returns a channel on which new ReplicationControllerDesires are sent,
+	// and a channel on which errors are reported.
+	// Send a value on quitAndAck to stop watching.
+	// watch() signals it has stopped by sending back a value on quitAndAck.
```

Comment:
It may not be necessary to send back a value on the quit channel to confirm that it has quit. Closing its output channels could be sufficient. I have no specific reason for sending back an ack on the quit channel other than that preparers do it too, but they probably had specific reasons for needing it whereas I... don't think I do.